### PR TITLE
[ntuple] Allow for vector commit of sealed pages

### LINKED ISSUE #10719: [ntuple] Enable page vector writes
### Explain what you would like to see improved
Currently, `RPageSink` provides the `CommitPage()` and `CommitSealedPage()` member functions.

OTOH, `RPageSinkBuf` provides buffering (and parallel compression) of pages and commits them one-by-one during a call to `CommitCluster()`.  In that case, pages are committed sequentially using the two aforementioned functions.  However, some backends, e.g. object stores such as DAOS, would benefit of a single call to write a vector of pages, which can then potentially be written in parallel.

### Optional: share how it could be improved
Add a `Commit{,Sealed}Pages()` member function to `RPageSink`; provide a default implementation that sequentially calls `CommitPage()`, but can be overridden in a subclass to implement parallel writing.

### PR DESCRIPTION
**Pull request by jalopezg-r00t**
This pull request extends `RPageSink` to allow for vector commit of sealed pages.  The prototype of the new functions is as follows:
```c++
virtual std::vector<RNTupleLocator> CommitSealedPageVImpl(std::span<RPageStorage::RSealedPageGroup> ranges);
void CommitSealedPageV(std::span<RPageStorage::RSealedPageGroup> ranges);
```

These changes enable (parallel) vector writes in concrete backends, e.g. in DAOS.  Derived classes that do not override the base implementation will see a sequence of calls to `CommitSealedPage()`.

The first measurements of vector writes on DAOS based on this PR, indicate an improvement of _an order of magnitude_ in the write throughput.

The refactor in this PR passes all the RPageSinkBuf tests.

## Changes or fixes:
- Add `CommitSealedPageV()` to the public interface of `RPageSink`. This function carries out a vector write of a number of sealed page ranges.  Each range applies to a given columnId.
The default implementation of `CommitSealedPageVImpl()` sequentially calls `CommitSealedPageImpl()` for each page on each range, but derived classes can override it to provide an optimized implementation.
- Use `CommitSealedPageV()` in RPageSinkBuf.  If a buffered column contains only sealed pages, commit the whole range
via `CommitSealedPageV()`.
- If all buffered columns consist solely of sealed pages, coalesce all pages to be committed in a single `CommitSealedPageV()` call that includes a sealed page range per column.

## Checklist:
- [X] tested changes locally
- [X] updated the docs (if necessary)

This PR fixes #10719.

### CODE REVIEW

**Review by jblomer — CHANGES_REQUESTED** [review]
Good work!  I have two name changes suggestions
  - `CommitSealedPages` --> `CommitSealedPageV` (same for the `Impl` method); this is in line with other vector signatures in RNTuple and I find the capital `V` easier to distinguish from the singular/scalar method call
  - `RSealedPageRange` --> `RSealedPageGroup`, since we call the set of pages of the same column in the same cluster a page group in the docs.

Other than that, let me have another look.  I have the feeling that the patch can still be simplified but I'm not yet sure about it.

File: tree/ntuple/v7/inc/ROOT/RPageSinkBuf.hxx [hunk 1]
```diff
@@ -81,18 +80,35 @@ private:
...
+      {
+         return fSealedPages.emplace(std::end(fSealedPages));
+      }
+
    private:
       RPageStorage::ColumnHandle_t fCol;
       // Using a deque guarantees that element iterators are never invalidated
       // by appends to the end of the iterator by BufferPage.
       std::deque<RPageZipItem> fBufferedPages;
+      // Pages that have been alredy sealed by a concurrent task. A vector commit can be issued if all
```
Comment (mxxo) [inline tree/ntuple/v7/inc/ROOT/RPageSinkBuf.hxx:109]:
```suggestion
      // Pages that have been already sealed by a concurrent task. A vector commit can be issued if all
```

**Review by mxxo — COMMENTED** [review]
Looks good to me!

**Comment by jalopezg-r00t** [comment]
> Looks good to me!

Thanks for reviewing, @mxxo!  During the first DAOS tests using the `ofi+tcp` provider, we measured over 2 GB/s for the default of 64 KiB pages (before it was ~200 MB/s).  :slightly_smiling_face:

File: tree/ntuple/v7/inc/ROOT/RPageStorage.hxx [hunk 2]
```diff
@@ -192,6 +206,13 @@ protected:
    virtual RNTupleLocator CommitPageImpl(ColumnHandle_t columnHandle, const RPage &page) = 0;
    virtual RNTupleLocator CommitSealedPageImpl(DescriptorId_t columnId,
                                                const RPageStorage::RSealedPage &sealedPage) = 0;
+   /// Vector commit of preprocessed pages. The `ranges` vector specifies a range of sealed pages to be
+   /// committed for each column.  The returned vector contains, in order, the RNTupleLocator for each
+   /// page on each range in `ranges`, i.e. the first N entries refer to the N pages in `ranges[0]`,
+   /// followed by M entries that refer to the M pages in `ranges[1]`, etc.
+   /// The default is to call `CommitSealedPageImpl` for each page; derived classes may provide an
+   /// optimized implementation though.
+   virtual std::vector<RNTupleLocator> CommitSealedPageVImpl(const std::vector<RPageStorage::RSealedPageGroup> &ranges);
```
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RPageStorage.hxx:215]: Should we rather use an `std::span` for the `ranges` input in this interface and in `CommitSealedPageV`?

File: tree/ntuple/v7/src/RPageSinkBuf.cxx [hunk 3]
```diff
@@ -92,10 +94,39 @@ ROOT::Experimental::Detail::RPageSinkBuf::CommitClusterImpl(ROOT::Experimental::
...
+
+      for (auto &bufColumn : fBufferedColumns)
+         bufColumn.DrainBufferedPages();
+      return fInnerSink->CommitCluster(nEntries);
+   }
+
+   // Otherwise, try to do it per column
    for (auto &bufColumn : fBufferedColumns) {
-      for (auto &bufPage : bufColumn.DrainBufferedPages()) {
+      auto hasSealedPagesOnly = bufColumn.HasSealedPagesOnly();
```
Comment (jblomer) [inline tree/ntuple/v7/src/RPageSinkBuf.cxx:116]: Let's make sure that we have unit tests for the three cases:
  - All pages sealed
  - All pages of a column sealed
  - Some pages sealed some not

**Review by jblomer — CHANGES_REQUESTED** [review]
I think it's essentially good to go, with some minor modifications.

Nice work!

**Review by jblomer — COMMENTED** [review]
The new unit tests do help but I meant something else: the buffered sink has quite a nice logic to try and make as much use as possible of vector writes.  However, I think that logic is not fully tested.  In practice, we have cases were only scalar writes are used and cases where everything is precompressed and we can submit one large vector write.  But the fine grained options in-between -- some columns in vector writes and some columns not -- are, I think, not tested.  I'm not even sure if such cases can occur in practice.
One option is to just add a comment on that possible optimization and to remove the more sophisticated parts of the vector write logic - i.e. either one big vector write or all scalar writes.

**Comment by jalopezg-r00t** [comment]
> But the fine grained options in-between -- some columns in vector writes and some columns not -- are, I think, not tested. I'm not even sure if such cases can occur in practice.
> One option is to just add a comment on that possible optimization and to remove the more sophisticated parts of the vector write logic - i.e. either one big vector write or all scalar writes.

Those cases could occur, but I think it is (probably) a misuse of the current API; I have updated the test cases to include that.  Let me know what you think; otherwise, as you said, we could just remove that possible per-column vector write optimization.

**Comment by jblomer** [comment]
> Those cases could occur, but I think it is probably a misuse of the API. I have updated the test cases to include that. Let me know what you think; otherwise, as you said, we could just remove that possible per-column vector write optimization.

That's a nice way to test it! Still, for reasons of maintainability I'd be more in favor of removing the per-column vector write optimization.  It's quite some code that is most likely never executed.

**Review by jblomer — APPROVED** [review]

**Pull request merged**